packages/controls/src/controls/definition.ts

```diff
@@ -0,0 +1,101 @@
+import { z } from 'zod'
+
+import { type ParseResult } from '../lib/zod'
+
+import { type Data } from '../common/types'
+import {
+  type CopyContext,
+  type MergeContext,
+  type MergeTranslatableDataContext,
+} from '../context'
+import { type IntrospectionTarget } from '../introspection'
+import {
+  Serializable,
+  serializeObject,
+  type SerializedRecord,
+} from '../serialization'
+
+import { ControlInstance, type SendMessage } from './instance'
+
+export type SchemaType<T> = z.ZodType<T>
+export type SchemaTypeAny = SchemaType<any> | z.ZodBranded<SchemaType<any>, any>
+
+export abstract class ControlDefinition<
+  ControlType extends string = string,
+  Config = unknown,
+  DataType extends Data = Data,
+  ValueType extends Data = Data,
+  ResolvedValueType = Data | unknown,
+  InstanceType extends ControlInstance<any> = ControlInstance<any>,
+> extends Serializable {
+  constructor(readonly config: Config) {
+    super()
+  }
+
+  abstract get controlType(): ControlType
+
+  abstract get schema(): {
+    definition: SchemaType<unknown>
+    type: SchemaType<ControlType>
+    data: SchemaType<DataType>
+    value: SchemaType<ValueType>
+    resolvedValue: SchemaType<ResolvedValueType>
+  }
```

Comment:
Exposing the schemas is crucial for implementing composable controls (list and shape).

Reply:
Yes, but only because this signature mandates defining all of these schemas on all of the controls. I think we can drop `value` and `resolvedValue` from the formal `ControlDefinition` interface and leave the decision whether it's beneficial to have these two schemas or not up to an individual control (even though we never use `value` and `resolvedValue` schemas to parse runtime data, oftentimes having them does simplify corresponding type definition for controls where both data and values depend on the config).

Reply:
I'm going to leave these intact for now, though, as a quick attempt at removing `value` and `resolvedValue` schemas broke `ResolvedValueType` type inference pretty much everywhere, and we still have more important issues to tackle in this cycle.